src/julia.jl

```diff
@@ -33,4 +33,139 @@ function multireshape(parent::AbstractArray, dimss::Array{Tuple{Vararg{Int}}})
     multireshape(parent, dimss...)
 end
 
+
+"""
+    pairwise(f::Function, a; kwargs...)
+
+Apply the function `f` to all pairwise combinations of elements from `a`. Set
+keyword `symmetric = true` to only calculate `f(x,y)` when `f(x,y) == f(y,x)`.
+"""
+function pairwise(f::Function, a; symmetric = false)
+    n = length(a)
+    f = tryfunc(f, a[1], a[1])
+    first = f(a[1], a[1])
+    r = Matrix{typeof(first)}(n, n)
+    pairwise!(f, a, r; symmetric = symmetric)
+    r
+end
+
+"""
+    pairwise(f::Function, a, output; kwargs...)
+
+Apply the function `f` to all pairwise combinations of elements from `a`, and
+return a `Vector` of the results if `output = :Vector`. The Vector form is
```

Comment:
I see.  But this is a very weird and un-Julian API.  You pass a symbol for the output argument, but it only accepts one symbol, and the symbol is only used to determine the number of arguments.

My first inclination is that these should be different functions, if this function is even worth defining at all.   But an alternative would be to pass the *type* `Vector` or `Matrix`, since the compiler *can* dispatch on `pairwise(f, a, ::Type{Vector})` vs `pairwise(f, a, ::Type{Matrix})`